nodeletctl/pkg/nodeletctl/nodeletctl.go

```diff
@@ -106,6 +106,130 @@ func CreateCluster(cfgPath string) error {
+	}
+
+	if clusterCfg.MasterVipEnabled {
+		if clusterCfg.MasterVipInterface == "" {
+			return fmt.Errorf("MasterVipInterface cannot be empty when MasterVip is enabled")
+		}
+		if clusterCfg.MasterVipVrouterId < 1 || clusterCfg.MasterVipVrouterId > 255 {
+			return fmt.Errorf("MasterVipVrouterId should be in the range 1-255 when MasterVip is enabled")
+		}
+	}
+
+	// TODO: Need to validate if nodelet version in the bootstrap is the supported version, possibly by
+	// checking a manifest file in the rpm/deb package
+
+	if err != nil {
+		return fmt.Errorf("failed to DetermineKubePkgName: %s", err)
+	}
+
+	globalClusterStatus = new(ClusterStatus)
+	globalClusterStatus.statusMap = make(map[string]*NodeStatus)
+
+	var masterList = make(map[string]string)
+	for _, host := range clusterCfg.MasterNodes {
+		if host.NodeIP != nil {
+			masterList[host.NodeName] = *host.NodeIP
+		} else {
+			masterList[host.NodeName] = host.NodeName
+		}
+	}
+
+	masters, err := GetCurrentMasters(clusterCfg)
+	if err != nil {
+		return fmt.Errorf("Failed to get active K8s masters: %s", err)
+	}
+
+	workers, err := GetCurrentWorkers(clusterCfg)
+	if err != nil {
+		return fmt.Errorf("Failed to get active K8s workers: %s", err)
+	}
+
+	if len(clusterCfg.MasterNodes) != len(masters) {
+		return fmt.Errorf("Masters in the cluster and Bootstrap config doesnt match", err)
+	}
+
+	if len(clusterCfg.WorkerNodes) != len(workers) {
+		return fmt.Errorf("Workers in the cluster and Bootstrap config doesnt match", err)
+	}
+
+	// Upgrade each master node sequentially
+	for _, host := range clusterCfg.MasterNodes {
+		zap.S().Infof("Upgrading master node: ", host.NodeName)
+		// Build nodeletconfig and deployer
+		nodeletCfg := new(NodeletConfig)
+		setNodeletClusterCfg(clusterCfg, nodeletCfg)
+		nodeletCfg.HostId = host.NodeName
+		if host.NodeIP != nil {
+			nodeletCfg.HostIp = *host.NodeIP
+		} else {
+			nodeletCfg.HostIp = host.NodeName
+		}
+		nodeletCfg.NodeletRole = "master"
+		nodeletCfg.MasterList = &masterList
+		nodeletCfg.EtcdClusterState = "new"
+
+		nodeletSrcFile, err := GenNodeletConfigLocal(nodeletCfg, masterNodeletConfigTmpl)
+		if err != nil {
+			zap.S().Infof("Failed to generate config: %s", err)
+			return fmt.Errorf("Failed to generate config: %s", err)
+		}
+		zap.S().Debugf("master nodeletsrc file %s", nodeletSrcFile)
+
+		deployer, err := GetNodeletDeployer(clusterCfg, globalClusterStatus, nodeletCfg, nodeletCfg.HostIp, nodeletSrcFile)
+		if err != nil {
+			zap.S().Errorf("failed to get nodelet deployer: %v", err)
+			return fmt.Errorf("failed to get nodelet deployer: %v", err)
+		}
+		globalClusterStatus.statusMap[host.NodeName] = &NodeStatus{
+			deployer: deployer,
+		}
+
+		if err := deployer.UpgradeNode(); err != nil {
+			SetClusterNodeStatus(deployer.clusterStatus, deployer.nodeletCfg.HostId, "failed", err)
+		}
+	}
+
+	// Upgrade each worker node sequentially
+	for _, host := range clusterCfg.WorkerNodes {
```

Comment:
Can we do this all in parallel like when we create or scale? Or in batches like qbert